### ISSUE TITLE
Random thoughts...

### ISSUE THREAD

**Issue by pfalcon** [MegaIng/python-magic#1]
... Can we please have this called "python-encoding-magic", and it to stick to that, because there're so many projects which try to embrace unembraceable and turn into little messy monsters, and very few which try to stick to one thing, and do it in clear, generic, reusable way...

... In that regard, would be possible to get fstringbackport.py out of encoding_magic package, and ideally into its own package, if not, then at least as a kind of "example" in the top of the repo? (But I guess the idea is to be able to ship it to users via pip, and then a separate package is the best option.) ...

**Comment by MegaIng** [comment]
No. This will be a collection of different magic ides by me. This isn't supposed to be a single package. You can pick out the stuff you want and put it together in a concise way. The idea isn't to ship it via pip (well, at least not in this repo. Maybe I will create a separate repo with only a small section.) This is actually kind hard for the encoding magic stuff since we don't have a nice general way of making a library preload stuff into the interpreter.

**Comment by MegaIng** [comment]
You are of course open to maintain a project with only the magic encoding stuff yourself. But I wont.

**Comment by pfalcon** [comment]
Fair enough, but you aren't (much) improving on anything, and just duplicating work done before you by other folks. E.g. this guy: https://github.com/aroberge/ideas/tree/master/ideas has similar stuff, which tries to mix up both import hooks and encoding hacks, infrastructure for them and actual syntax transformers, etc.

> This is actually kind hard for the encoding magic stuff since we don't have a nice general way of making a library preload stuff into the interpreter.

Of course we have such a way: `python3 -m my_preloader <script> <script_args>`.

> You are of course open to maintain a project with only the magic encoding stuff yourself. But I wont.

I personally consider encoding hacks a dead-end. I'll be making my generic import-hook based thingy, yeah. Initial discussion is here: https://mail.python.org/archives/list/python-ideas@python.org/thread/NIQQVA3OJFUHL3INBMBUBMTP24W74XEO/ . Consider posting info about your stuff on python-ideas, it's such a boring place otherwise ;-).

**Comment by MegaIng** [comment]
> duplicating work done before you by other folks

I know, and I don't care.

> Of course we have such a way: python3 -m my_preloader <script> <script_args>.

Requiring you too always execute scripts like that, which can be avoided by using a `.pth` or `sitecustomize.py` file. Sadly, there is no reliably way of installing this with pip/distutils/setuptools (at least that I could find).

**Comment by pfalcon** [comment]
> I know, and I don't care.

And as you guessed, I do ;-). Someone should start to care some day.

> Requiring you too always execute scripts like that

So what, you always need to execute `python3 <script>` anyway. No, wait, you don't always need to, you can add a shebang and run it directly! But then you either can make my_app_loader.py which will contain just `install_hooks(); run_actual_app()`, and run that. The Python even has that pattern encoded in its module system - when you run a package (and any application beyond a trivial will be a package, not a single script), what actually gets run is `__main__` module in that package, and it would contain those `install_hooks(); run_actual_app()`.

**Comment by MegaIng** [comment]
> And as you guessed, I do ;-). Someone should start to care some day.

The fix it. This repo isn't supposed to be a public project. I am just going to collect different magic approaches here.

> shebang

Doesn't work on windows without a bit more trouble. Also can't official pass additional parameters to python.

> So what, you always need to execute python3 <script> anyway

Unless of course some other process executes `python <script>` in a context where you can't easily change it (e.g. subprocess call, IDE, unit testing, external platforms, legacy code, ...).

**Comment by pfalcon** [comment]
> The[n] fix it.

I on my own won't be able. Only "we" (people who're interested in more advanced usage of Python than the committee wants us to) would be able. If that "we" exists at all, and it's not just disparate dudes sitting each in his own hole, digging it with a thimble.

And to understand if there's any "we", we need to talk to each other. In that regard, it was a nice conversation for starters, thanks!

**Comment by MegaIng** [comment]
If I find actually usefull tricks that I also want to maintain, I will: https://github.com/MegaIng/pattern-matching

This is more collection of a lot of different ideas. I wouldn't consider encoding hacks in the current form a good idea. The current implementation is more of a quick way of testing ideas.

**Comment by pfalcon** [comment]
Yeah, seeing repos like that, is how I started to follow you on github. I agree there's usually not much opportunity for direct cooperation, as everyone has different sub-interests and priorities, but I find it really helpful to be in loop on what other people work on (even if for "Ok, that guy works on that, so that area is covered. I then work on another thing I had in mind for a while.")

Pattern matching is definitely interesting, and I follow what they want to add to the CPython mainline, concerned that it wasn't "too bad" comparing to alternative variants. (Note that I maintain my own small Python dialect, https://github.com/pfalcon/pycopy , so all my direct interests go thru the prism of it, though again, I try follow the scene in general.)

**Comment by MegaIng** [comment]
Aha, you didn't come here via [stackoverflow](https://stackoverflow.com/questions/65296925/f-string-debugging-shorthand-in-python-3-6)?

**Comment by pfalcon** [comment]
Nope, just following a few people who do advanced Python hacking, and saw a new repo in my github activity feed. I have to admit I'm not much interested in f-strings myself - I use Python since 1.5.2, and never felt a need for them ;-). But I feel a strong need for a generic framework which would allow to install, select on a per-application basis, and exchange with other people the syntax/semantics transformers for Python, because different people have so wide and different ideas, that we should just embrace the idea of "Python dialects", and should make it easy for anyone to try any of them.

What prompted me to actually write was seeing:

```
# -*- coding: rewrite-{rewriter-names} -*-
```

That's just how it should be done - a user should have ability to compose different transformers together, all under auspices of a singe, simple, generic tool.

Again, that's what I want to do myself, just using import hooks (I have additional constraint that it should work with *both* my Pycopy and CPython).

**Comment by MegaIng** [comment]
> I have additional constraint that it should work with both my Pycopy and CPython

Funny thing in this regard: What I am doing here is completely within the python specs. If this doesn't work on an implementation, the implementation is wrong. (I think this applies even to python2.7) (This is also the reason I create such a general system). Only problem is a reliable way of installing this. But you are of course correct that using a wrapper is a decent option.

**Comment by pfalcon** [comment]
Pycopy is a minimalist subset, and intended to be so. It's firmly in 21th century with being utf-8 only. There's no support for program source character encodings, and no reason to add it. So, I'm definitely biased saying that encoding hacks is dead-end ;-).
